Issue 1183: Migration - Teamspaces

dleszcz opened the pull request:
- Teamspaces list view
- Add/Edit & Delete model dialogs
- Model settings view
- Upload model file dialog
- Model revisions dialog
- Download model action
- Attach all models actions using sagas
- Implement design based on https://xd.adobe.com/view/9180ddc9-5807-4b95-6985-2f573e452411-e7b1/screen/62981225-547c-48f3-93c7-686ac214f640/
- Moved styles to styled components

simplypixi commented on frontend/routes/modelSettings/modelSettings.component.tsx:
```diff
@@ -0,0 +1,390 @@
+
+		if (!prevProps.modelSettings.properties && properties) {
+			if (topicTypes && topicTypes.length !== prevState.topicTypes.length) {
+				this.setState({ topicTypes });
```
Each `.setState` invokes async state change. I think it's better to collect these changes and then call `.setState` once.

```suggestion
const changes = {};

if(something) {
   changes.foo = 'bar';
}

if (!isEmpty(changes)){
   this.setState(changes);
}
```

simplypixi commented on frontend/routes/modelSettings/modelSettings.component.tsx:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	public componentDidUpdate(prevProps, prevState) {
+		const properties = this.props.modelSettings.properties;
```
```suggestion
		const { properties, topicTypes, surveyPoints, elevation, angleFromNorth }  = this.props.modelSettings;
```
You can easily desctructure the object's props. It's prettier and clearier. ;)

simplypixi commented on frontend/routes/modelSettings/modelSettings.component.tsx:
```diff
@@ -0,0 +1,390 @@
+
+		if (prevSurveyPoints !== surveyPoints && surveyPoints.length) {
+			const [ { latLong, position } ] = surveyPoints;
+			if (this.state.axisX !== position[0]) {
```
```const {axisX, axisY, axisZ, latitude, longitude } = this.state;```

simplypixi commented on frontend/routes/modelSettings/modelSettings.component.tsx:
```diff
@@ -0,0 +1,390 @@
+	}
+};
+
+const ENTER_KEY = 'Enter';
```
According to our conversation, you could move `ENTER_KEY` and `unitsMap` to new constants' files.

simplypixi commented on frontend/routes/modelSettings/modelSettings.component.tsx:
```diff
@@ -0,0 +1,390 @@
+						<Headline color="primary" variant="subheading">Model Information</Headline>
+						<Grid>
+							<FieldsRow container wrap="nowrap">
+								<Field name="id" render={({ field, form }) => (
```
Please check whether all occurrences of `form` params are necessary. ;)

simplypixi commented on frontend/routes/teamspaces/components/modelDialog/modelDialog.styles.ts:
```diff
@@ -0,0 +1,87 @@
+
+import styled from 'styled-components';
+import { Field } from 'formik';
+import { CustomTable } from '../../../components/customTable/customTable.component';
```
You can merge these two imports

simplypixi commented on frontend/routes/teamspaces/components/modelDialog/modelDialog.styles.ts:
```diff
@@ -0,0 +1,87 @@
+	min-width: 480px;
+
+	${FieldWrapper} {
+    margin-right: 10px;
```
wrong indentation

simplypixi commented on frontend/routes/teamspaces/components/modelDialog/modelDialog.styles.ts:
```diff
@@ -0,0 +1,87 @@
+
+	> div {
+		width: 100%;
+		border-bottom: 1px solid rgba(0,0,0,.06);
```
`COLORS.BLACK_6`

simplypixi commented on frontend/routes/teamspaces/components/modelDialog/modelDialog.styles.ts:
```diff
@@ -0,0 +1,87 @@
+
+		&:first-child {
+			&:first-child {
+				border-right: 1px solid rgba(0,0,0,.06);
```
`COLORS.BLACK_6`

simplypixi commented on frontend/routes/teamspaces/components/uploadModelFileDialog/uploadModelFileDialog.component.tsx:
```diff
@@ -0,0 +1,180 @@
+	StyledDialogActions
+} from './uploadModelFileDialog.styles';
+
+const acceptedFormat = [
```
You should move this variable to the new constants file

simplypixi commented on frontend/routes/teamspaces/components/uploadModelFileDialog/uploadModelFileDialog.component.tsx:
```diff
@@ -0,0 +1,180 @@
+
+	public onInputChange = (event) => {
+		this.setState({
+			file: this.inputFileRef.current.files[0].name
```
Why not `event.target.files[0].name`?

dleszcz replied:
Oh, this is really better. I'll change it

simplypixi commented on frontend/services/api/teamspace.ts:
```diff
@@ -33,3 +33,30 @@ export const fetchTeamspace = (teamspace) => {
+ * @param teamspace
+ * @param modelId
+ */
+export const getModelSettings = (teamspace, modelId) => {
```
I think that these three functions should be moved `api/models.ts`.

simplypixi commented:
@carmenfan I reviewed the code. I found some bugs but I think you can start testing this PR.

dleszcz commented on frontend/components/account/js/account-items.component.ts:
```diff
@@ -113,7 +113,7 @@ class AccountItemsController implements ng.IController {
 
 		this.showProgress = true;
-		this.modelTypes = ["Architectural", "Structural", "Mechanical", "GIS", "Other"];
+		this.modelTypes = ["Structural", "Mechanical", "GIS", "Other"];
```
Sorry, my bad. I'll fix it now.

dleszcz commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,101 @@
+};
+
+// Models
+const updateModelSuccess = (state = INITIAL_STATE, action) => {
```
fixed

dleszcz commented:
@carmenfan PR is ready

sanmont3drepo commented on frontend/routes/components/dateTime/dateTime.styles.ts:
```diff
@@ -0,0 +1,20 @@
+/**
```
Is this file necessary?

dleszcz replied:
You're right. I removed these generated by plop styles files.

sanmont3drepo commented on frontend/routes/components/dialogContainer/components/confirmDialog/confirmDialog.styles.ts:
```diff
@@ -0,0 +1,20 @@
+/**
```
Again if there is no style defined, should this file exist?

sanmont3drepo commented on frontend/routes/components/panel/panel.component.tsx:
```diff
@@ -24,16 +24,19 @@ import { MuiTheme, theme } from '../../../styles';
+	title?: string;
 	children: any;
 	paperProps?: any;
+	renderCustomTitle?: any;
```
I would have gone with only "title" without the renderCustomTitle that could be a string or a JSX.Element . This is only to take in consideration, not needed for the pr.

simplypixi replied:
Good point! ;)

sanmont3drepo commented on frontend/routes/modelSettings/modelSettings.styles.ts:
```diff
@@ -0,0 +1,102 @@
+	margin: 30px 0 15px;
+`;
+
+export const StyledChip = styled(Chip)`
```
Later if we are going to have another component that uses chips this should be in a more general styles file , not particular for this component. Preferably instead of a styled component, in the theme. Every chip in the app should look the same.

carmenfan commented:
@simplypixi 
![image](https://user-images.githubusercontent.com/11945337/47921387-6250be00-deac-11e8-95f4-b09b91bac29a.png)
`yarn run build` is now taking quite a long time... (was ~1min pre react, ~2mins before this issue). Anything we can do about it?

simplypixi commented:
> @simplypixi
> ![image](https://user-images.githubusercontent.com/11945337/47921387-6250be00-deac-11e8-95f4-b09b91bac29a.png)
> `yarn run build` is now taking quite a long time... (was ~1min pre react, ~2mins before this issue). Anything we can do about it?

@carmenfan
The build should take more time (because of two libraries on the board), but the difference shouldn't so big. I wanted to speed up the build after migration, but in this situation, I will have to do it earlier. Will you mind if I take care of it now (as much as the current stack allows)? It should take me about at most two working days, but because of that, the migration of pages could be a bit delayed.

carmenfan commented:
> > @simplypixi
> > ![image](https://user-images.githubusercontent.com/11945337/47921387-6250be00-deac-11e8-95f4-b09b91bac29a.png)
> > `yarn run build` is now taking quite a long time... (was ~1min pre react, ~2mins before this issue). Anything we can do about it?
> 
> @carmenfan
> The build should take more time (because of two libraries on the board), but the difference shouldn't so big. I wanted to speed up the build after migration, but in this situation, I will have to do it earlier. Will you mind if I take care of it now (as much as the current stack allows)? It should take me about at most two working days, but because of that, the migration of pages could be a bit delayed.

@simplypixi 
If it's easier to do it after migration then we could leave it after. Up to you.

simplypixi commented:
@dleszcz 

I think you should remove (disabled) horizontal scroll on "Model settings" and add simplebar to "Teamspaces" and "Model settings".

> This cancel button doesn't work 😆 :
And the padding of dialog content has to be smaller

simplypixi commented:
> > > @simplypixi
> > > ![image](https://user-images.githubusercontent.com/11945337/47921387-6250be00-deac-11e8-95f4-b09b91bac29a.png)
> > > `yarn run build` is now taking quite a long time... (was ~1min pre react, ~2mins before this issue). Anything we can do about it?
> > 
> > 
> > @carmenfan
> > The build should take more time (because of two libraries on the board), but the difference shouldn't so big. I wanted to speed up the build after migration, but in this situation, I will have to do it earlier. Will you mind if I take care of it now (as much as the current stack allows)? It should take me about at most two working days, but because of that, the migration of pages could be a bit delayed.
> 
> @simplypixi
> If it's easier to do it after migration then we could leave it after. Up to you.

Yes, it's definitely easier to speed up build after migration. AngularJS limits some optimization possibilities. But I'll prepare some tools to control the size and time of the build during migration.

simplypixi commented on frontend/components/account/js/account-items.component.ts:
```diff
@@ -113,7 +113,7 @@ class AccountItemsController implements ng.IController {
 		this.ViewerService.reset();
```
This file isn't used anymore

simplypixi commented on frontend/constants/model-parameters.ts:
```diff
@@ -0,0 +1,17 @@
+export const acceptedFileFormats = [
```
This file should have the name with camelCase syntax

simplypixi replied:
And constants should be defined using upper-case.

simplypixi commented on frontend/routes/teamspaces/components/uploadModelFileDialog/uploadModelFileDialog.component.tsx:
```diff
@@ -0,0 +1,185 @@
+		});
+	}
+
+	public getAcceptedFormats = () => acceptedFileFormats.map((format) => `.${format}`).toString();
```
I suppose that you have to get these formats from ClientConfigService.

simplypixi commented on frontend/modules/billing/billing.sagas.ts:
```diff
@@ -18,14 +18,11 @@
-import { DIALOG_TYPES } from '../dialog/dialog.redux';
+import {
+	LoadingDialog
+} from './../../routes/components/dialogContainer/components/loadingDialog/loadingDialog.component';
```
```suggestion
} from '../../routes/components/dialogContainer/components';
```

simplypixi commented on frontend/modules/model/model.helpers.ts:
```diff
@@ -0,0 +1,15 @@
+};
+
+export const MODEL_STATUSES_MAP = {
+	queued: 'queued',
```
camelCase

simplypixi commented on frontend/modules/model/model.redux.ts:
```diff
@@ -0,0 +1,56 @@
+};
+
+const setPendingState = (state = INITIAL_STATE, { pendingState }) => {
+	return Object.assign({}, state, { isPending: pendingState });
```
```suggestion
	return { ...state, isPending: pendingState };
```

simplypixi commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -0,0 +1,132 @@
+		const { data: settings } = yield API.getModelSettings(teamspace, modelId);
+
+		yield all([
+			put(ModelActions.fetchSettingsSuccess(settings)),
```
I think these actions should be called one by one.

simplypixi commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -0,0 +1,132 @@
+
+export function* updateSettings({ teamspace, modelId, settings }) {
+	try {
+		const response = yield API.editModelSettings(teamspace, modelId, settings);
```
Shouldn't the store be updated?

simplypixi commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -0,0 +1,132 @@
+}
+
+export function* subscribeOnStatusChange({ teamspace, project, modelId }) {
+	const notificationService = yield getAngularService("NotificationService");
```
single quote

simplypixi commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,137 @@
+const updateModelSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.modelData.projectName);
```
```suggestion
	const projectIndex = projects.findIndex((project) => project.name === action.modelData.projectName);
	const foundProject = projects[projectIndex];
```

simplypixi commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,137 @@
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.modelData.projectName);
+	const projectIndex = projects.indexOf(findedProject);
+	const findedModel = findedProject.models.find((model) => model.name === action.modelData.name);
```
```suggestion
	const findedModel = findedProject.models.find((model) => model.name === action.modelData.name);
```
```suggestion
	const modelIndex = foundProject.models.findIndex((model) => model.name === action.modelData.name);
```
You dont need `foundModel`

simplypixi commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,137 @@
+const createModelSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.modelData.projectName);
```
same here

simplypixi commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,137 @@
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.modelData.projectName);
+	const projectIndex = projects.indexOf(findedProject);
```
and here

simplypixi commented on frontend/modules/teamspaces/teamspaces.redux.ts:
```diff
@@ -0,0 +1,137 @@
+const setModelUploadStatus = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.project);
```
and here

I think you can move this logic to separate method. ;)

dleszcz commented:
Hello @carmenfan! Can you test this PR again?

carmenfan commented:
![image](https://user-images.githubusercontent.com/11945337/48345840-17376780-e671-11e8-8289-3ad55cd6d54a.png)
Units should be disabled and unable to change on edit settings page

carmenfan commented:
For the revision menu, can we do it in this format please?
![image](https://user-images.githubusercontent.com/11945337/48353610-c2ebb200-e687-11e8-898c-abc04506b3e2.png)

main changes are:
- ditch the ID and put in the revision description
- Add time on timestamp display
- tool tip on hover that says `open revision`
- If there is no tag name, it should say `(empty)` instead of putting in the date again

:)

carmenfan commented:
Some more feedback from Pavol and our business team:
- Can we put the save/cancel buttons on a footer to ensure it is always visible? (this applies to all menus)
![image](https://user-images.githubusercontent.com/11945337/48357738-2d552000-e691-11e8-837b-894041a30d4b.png)

- New model/federations doesn't appear on newly created projects (require reload)

- Can we make the pop up headers to have the same height as the other ones?
![image](https://user-images.githubusercontent.com/11945337/48357882-845af500-e691-11e8-893d-5ef52235d80a.png)

- Can we capitalise the "t" to `My Teamspace`?
![image](https://user-images.githubusercontent.com/11945337/48357976-c08e5580-e691-11e8-8188-ef22c8395cbc.png)

simplypixi commented:
> > > Can we make the pop up headers to have the same height as the other ones?
> > 
> > 
> > Which spacing do you prefer? Bigger or smaller?
> > Personally, I think the bigger spacing is prettier. :)
> 
> Pavol says he wants it narrower.. 😆

Ok!

> main changes are:
> 
> * ditch the ID and put in the revision description
> * Add time on timestamp display
> * tool tip on hover that says `open revision`
> * If there is no tag name, it should say `(empty)` instead of putting in the date again

Is the tooltip necessary? These elements are clickable and have pointer cursor, so easy to understand that when I click any of these items I'll be redirected to it. What do you think about it?

carmenfan commented:
@simplypixi 
Maybe leave it for now, and I'll get Pavol to try the revised version and see what he thinks :)

simplypixi commented:
> @simplypixi
> Maybe leave it for now, and I'll get Pavol to try the revised version and see what he thinks :)

Sure, no problem!

dleszcz commented:
@carmenfan I have a a question to updating statuses of model revisions.

Currently on our local machines we can't upload model file (we have infinite status "queued" after each attempt to upload) for both Angular and React implementations. 
Do we need to install/configure something extra to make it work?

Other your requests are clear and fixed.

carmenfan commented:
> @carmenfan I have a a question to updating statuses of model revisions.
> 
> Currently on our local machines we can't upload model file (we have infinite status "queued" after each attempt to upload) for both Angular and React implementations.
> Do we need to install/configure something extra to make it work?
> 
> Other your requests are clear and fixed.

Yes, it's a bit complicated to make it work fully. The web server is only one box within the 3drepo infrastructure. Model uploads are done by another box with a C++ library(3drepobouncer) and a unity instance, facilitated by the queue.

So to test this fully you will have to setup the C++ box, or we would have to setup something on the cloud for you to connect.

But for the sake of what you are doing it may be easier to just test it with a dummy consumer on the other end.  I've created a nodeJS script that will imitate and spew out success messages upon receiving a message from the queue, you should be able to use this to test: https://www.dropbox.com/s/hsuch702km0xl0j/bouncerFaker.zip?dl=0

simplypixi commented:
@carmenfan
I pushed to this branch new linter config (as I promised on the other PR).
Moreover, I've added `stylelint` to lint styled-component's files. Currently, I'm not able to set all rules properly because of tslint capabilities and legacy code. tslint doesn't support rules for specific files e.g old Angular components.

I've updated scripts in `package.json` too. Now, `yarn lint` runs two linters (ts + styled-components).

Please, tell the team about it. ;)

dleszcz commented:
> > @carmenfan I have a a question to updating statuses of model revisions.
> > Currently on our local machines we can't upload model file (we have infinite status "queued" after each attempt to upload) for both Angular and React implementations.
> > Do we need to install/configure something extra to make it work?
> > Other your requests are clear and fixed.
> 
> Yes, it's a bit complicated to make it work fully. The web server is only one box within the 3drepo infrastructure. Model uploads are done by another box with a C++ library(3drepobouncer) and a unity instance, facilitated by the queue.
> 
> So to test this fully you will have to setup the C++ box, or we would have to setup something on the cloud for you to connect.
> 
> But for the sake of what you are doing it may be easier to just test it with a dummy consumer on the other end. I've created a nodeJS script that will imitate and spew out success messages upon receiving a message from the queue, you should be able to use this to test: https://www.dropbox.com/s/hsuch702km0xl0j/bouncerFaker.zip?dl=0

I tried to run a bouncerFaker, but I get this error:
![screenshot 2018-11-14 at 11 52 31](https://user-images.githubusercontent.com/17221890/48478181-cefc7e80-e803-11e8-9fb7-b88c53c85a08.png)

Do you know what is the reason? Model JSON files and folders are created in sharedSpace folder:
<img width="499" alt="screenshot 2018-11-14 at 11 57 33" src="https://user-images.githubusercontent.com/17221890/48478406-80031900-e804-11e8-9dcd-b50ba2156b7b.png"> 

My local config.js:
`cn_queue: {
    host: 'amqp://localhost:5672', //host/port to connect to the rabbitmq instance
    worker_queue: 'jobq', //Name of the work queue
    model_queue: 'modelq', //Name of the model queue
    callback_queue: 'callbackq', //Name of the callback queue
    upload_dir: 'uploads', //A folder to temporarily stick the uploaded file into
    shared_storage: "sharedSpace", //A folder that is shared between the compute nodes and web instance
    event_exchange: "eventExchange" //Name of the socket.io queue
  },`

carmenfan commented:
@dleszcz 
That's because it is a relative path configured on the 3drepo.io side, so when your fake bouncer runs (from a completely different directory), it's not going to find the path via relative path.

It should probably fix itself if you put the absolute path in your 3drepo.io config.

dleszcz commented:
> @dleszcz
> That's because it is a relative path configured on the 3drepo.io side, so when your fake bouncer runs (from a completely different directory), it's not going to find the path via relative path.
> 
> It should probably fix itself if you put the absolute path in your 3drepo.io config.

Thank you! Can you check your requests?

carmenfan commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -72,8 +72,15 @@ export function* downloadModel({ teamspace, modelId }) {
 
-export function* onModelStatusChanged({ modelData: { status }, teamspace, project, modelId }) {
-	yield put(TeamspacesActions.setModelUploadStatus(teamspace, project, modelId, modelStatusesMap[status]));
+export function* onModelStatusChanged({ modelData, teamspace, project, modelId }) {
```
I don't see this triggered, snackbar doesn't activate when the upload finishes...

Also, it would be nice to see the model name on the snackbar instead of just saying "model" as it can take a while before this happens.

carmenfan replied:
@dleszcz  can we do this?

carmenfan commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -72,8 +72,15 @@ export function* downloadModel({ teamspace, modelId }) {
+		yield put(SnackbarActions.show('Model uploaded succesfully'));
+	}
+	if (status === uploadFileStatuses.failed) {
+		yield put(SnackbarActions.show('Failed to import model'));
```
We need to display the error reason somewhere. So if we replace what was the message displayed on the model row to here, we need to display the error message here.

carmenfan commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -101,14 +108,14 @@ export function* uploadModelFile({ teamspace, project, modelId, fileData }) {
 			if (data.hasOwnProperty('errorReason') && data.errorReason.message) {
 				yield put(SnackbarActions.show(data.errorReason.message));
 			} else {
 				yield put(SnackbarActions.show('Model uploaded succesfully'));
```
typo: successfully

carmenfan commented on frontend/modules/model/model.sagas.ts:
```diff
@@ -101,14 +108,14 @@ export function* uploadModelFile({ teamspace, project, modelId, fileData }) {
 		const { data: { status }, data } = yield API.uploadModelFile(teamspace, modelId, formData);
 
-		if (status === uploadFileStatuses.OK) {
+		if (status === uploadFileStatuses.ok) {
```
I don't think this ever runs. I don't see a snackbar upon upload API request.

I think the status it returns is "upload", if it errored it'll always end up in the catch clause.

carmenfan commented:
@simplypixi 
> > It keeps refreshing when I log in, and eventually end up with session expired (some times 1 refresh, sometimes 2):
> 
> Did you do some specific steps? I can't reproduce this bug. :(

I managed to reproduce this now! There's something weird going on with the auto redirect when you log out on another browser tab.
To reproduce:
1. log in on a tab (let's call this tab A).
2. start a new tab (tab B) and go to the website.
3. log out on tab B - look at tab A, it should've auto logout
4. log in again on Tab A - it keeps refreshing and logging me out again!

![before](https://user-images.githubusercontent.com/11945337/48556331-98923280-e8db-11e8-80cd-262f8af3a383.gif)

It is really annoying and only way to fix this is by a clear cache and hard reload :(

simplypixi commented:
> @simplypixi
> 
> > > It keeps refreshing when I log in, and eventually end up with session expired (some times 1 refresh, sometimes 2):
> > 
> > 
> > Did you do some specific steps? I can't reproduce this bug. :(
> 
> I managed to reproduce this now! There's something weird going on with the auto redirect when you log out on another browser tab.
> To reproduce:
> 
> 1. log in on a tab (let's call this tab A).
> 2. start a new tab (tab B) and go to the website.
> 3. log out on tab B - look at tab A, it should've auto logout
> 4. log in again on Tab A - it keeps refreshing and logging me out again!
> 
> ![before](https://user-images.githubusercontent.com/11945337/48556331-98923280-e8db-11e8-80cd-262f8af3a383.gif)
> 
> It is really annoying and only way to fix this is by a clear cache and hard reload :(

Ok! ;)

carmenfan commented:
@simplypixi yep seems to work now 🎉 

Pavol found a bug (not related to your latest commit, he was testing on it yesterday), it looks like registerVerify is no longer getting the query string variables properly
![image](https://user-images.githubusercontent.com/11945337/48614629-989b3c80-e986-11e8-8561-278ff3bcb9ba.png)

dleszcz commented:
@carmenfan 
We with @simplypixi tried found reason of bugs related to setting model statuses. We've gone through all possible error paths, but we cannot find reason. 

Maybe you've changed something in the chat? That would explain why it doesn't work now. We subscribe on proper channel, but for some reason callback in method modelStatusChanged is not called. We are sure that 2 days ago this works fine. (FYI: files are uploaded to sharedSpace directory  correctly)

simplypixi commented:
> @simplypixi yep seems to work now 🎉
> 
> Pavol found a bug (not related to your latest commit, he was testing on it yesterday), it looks like registerVerify is no longer getting the query string variables properly
> ![image](https://user-images.githubusercontent.com/11945337/48614629-989b3c80-e986-11e8-8561-278ff3bcb9ba.png)
> 
> EDIT: session expire message doesn't seem to appear anymore

@carmenfan I'll check that ;)

simplypixi commented:
> > @simplypixi yep seems to work now 🎉
> > Pavol found a bug (not related to your latest commit, he was testing on it yesterday), it looks like registerVerify is no longer getting the query string variables properly
> > ![image](https://user-images.githubusercontent.com/11945337/48614629-989b3c80-e986-11e8-8561-278ff3bcb9ba.png)
> > EDIT: session expire message doesn't seem to appear anymore
> 
> @carmenfan I'll check that ;)

Fixed ;)

dleszcz commented:
@carmenfan Your last requests have been fixed today (statuses, timestamps, etc.)

simplypixi commented:
> The indefinite redirect bug came back.... @_@
This is weird...

> EDIT: session expire message doesn't seem to appear anymore
I've probably omitted that... Sorry about that.

@carmenfan I have a small proposal. Could we check and merge rewritten user session (from another PR) and login page to this branch? I know, this PR is getting bigger and bigger but now I'll have to fix some bugs for parts of the code, which is totally removed on my second PR. What do you think about it?

carmenfan commented:
> @carmenfan I have a small proposal. Could we check and merge rewritten user session (from another PR) and login page to this branch? I know, this PR is getting bigger and bigger but now I'll have to fix some bugs for part of the code, which is totally removed on my second PR. What do you think about it?

that's fine yep, we'll try to look into it today

carmenfan commented:
A few more from pavol:

1. after a user registers, the message that tells them to check the email is missing, it goes straight to log in screen with an error saying the user is not logged in

2. can we remember the state of the teamspace collapse/expand when the user comes back from the viewer?

dleszcz commented:
@carmenfan  Referring to timestamps you are absolutely right, sorry this is my mistake in updating model function. 
But immediately after successfully upload model we have to generate timestamp on frontend because we don't have any timestamp in response (the same was done in the AngularJS application - _screenshot attached_). Of course after refreshing page the model with timestamp is fetching from backend.
<img width="924" alt="timestampngfrontend" src="https://user-images.githubusercontent.com/17221890/48781438-4cc8fa00-ecdc-11e8-88eb-b9f472d12793.png">

carmenfan commented:
@dleszcz actually looking at the backend code, you should have the timestamp from modelData?

![image](https://user-images.githubusercontent.com/11945337/48836423-9f58f380-ed7a-11e8-9b7e-0db9b77a94b4.png)
This is what I got from `modelData` in `onModelStatusChanged` of `model.saga.ts`

carmenfan commented:
@dleszcz  this is because you used a fake bouncer. You can assume this will exist if it's not fake! :)

dleszcz commented:
@carmenfan ok, it makes sense ;)

dleszcz commented:
@carmenfan What about creating/editing federation? I think then backend also should return timestamp in response.data object.

carmenfan commented:
> 
> 
> @carmenfan What about creating/editing federation? I think then backend also should return timestamp in response.data object.

indeed. What it really should do is utilise the queue. I've made an issue about it a while back and it's probably out of scope for you guys. So for now I've modified the route on the backend and it should now return the settings, which should have a timestamp upon success.

dleszcz commented:
> > @carmenfan What about creating/editing federation? I think then backend also should return timestamp in response.data object.
> 
> indeed. What it really should do is utilise the queue. I've made an issue about it a while back and it's probably out of scope for you guys. So for now I've modified the route on the backend and it should now return the settings, which should have a timestamp upon success.

Thanks! Now backend returns settings on update, but NOT on create new one. It shouldn't? 
I fixed this in this commit:
https://github.com/3drepo/3drepo.io/pull/1224/commits/0eaf4f2a4bfadc4d76f7c02d0d0bb448a999bb4d
But let me know if it is not correct.

Also improved last requests, so if I can ask you for a feedback it would be great.

dleszcz commented:
FYI: We found one more issue related with switching between viewer and teamspaces.

dleszcz commented:
> A few more from pavol:
> 
> 1. after a user registers, the message that tells them to check the email is missing, it goes straight to log in screen with an error saying the user is not logged in
> 2. can we remember the state of the teamspace collapse/expand when the user comes back from the viewer?

These things are also done.

simplypixi commented:
> FYI: We found one more issue related with switching between viewer and teamspaces.

@carmenfan Fixed!

carmenfan commented:
> > > @carmenfan What about creating/editing federation? I think then backend also should return timestamp in response.data object.
> > 
> > 
> > indeed. What it really should do is utilise the queue. I've made an issue about it a while back and it's probably out of scope for you guys. So for now I've modified the route on the backend and it should now return the settings, which should have a timestamp upon success.
> 
> Thanks! Now backend returns settings on update, but NOT on create new one. It shouldn't?
> I fixed this in this commit:
> [0eaf4f2](https://github.com/3drepo/3drepo.io/commit/0eaf4f2a4bfadc4d76f7c02d0d0bb448a999bb4d)
> But let me know if it is not correct.
> 
> Also improved last requests, so if I can ask you for a feedback it would be great.

yes it should! Thanks for comitting it!

Charence commented:
![screenshot 2018-11-23 14 04 45](https://user-images.githubusercontent.com/1438332/48950355-c528f880-ef32-11e8-8c86-cddbe6a3ce39.png)

@dleszcz @simplypixi I seem to be having problems running this branch (please see screenshot). I have done `yarn install` and `yarn run build`. I've tried on Chrome 70 and Firefox 63 on OS X 10.12.6. Any ideas? Thanks!

simplypixi commented:
> ![screenshot 2018-11-23 14 04 45](https://user-images.githubusercontent.com/1438332/48950355-c528f880-ef32-11e8-8c86-cddbe6a3ce39.png)
> 
> @dleszcz @simplypixi I seem to be having problems running this branch (please see screenshot). I have done `yarn install` and `yarn run build`. I've tried on Chrome 70 and Firefox 63 on OS X 10.12.6. Any ideas? Thanks!

Hi @Charence !
Are you sure that you ran `build and not `watch`? I see `Development Environment` in your dev tools. If it's the build, please try to clean yarn cache. Then remove node_modules and call yarn commands again.
Maybe it's a silly question, but do you pulled the latest changes? Which commit is the latest on your machine?

Small tip: You can you use `yarn` and `yarn build`. Yarn doesn't need `run` command.

simplypixi commented:
> > Hi @Charence !
> > Are you sure that you ran `build and not `watch`? I see `Development Environment` in your dev tools. If it's the build, please try to clean yarn cache. Then remove node_modules and call yarn commands again.
> > Maybe it's a silly question, but do you pulled the latest changes? Which commit is the latest on your machine?
> > Small tip: You can you use `yarn` and `yarn build`. Yarn doesn't need `run` command.
> 
> Thanks for the tip!
> 
> Unfortunately, I'm still getting the error.
> 
> It says `Development Environment` because I have set `development: true` in my backend configuration. I tried to remove this.
> 
> I also did a `yarn cache clean`, remove node_modules/ from frontend (and backend), remove public/, ran `yarn`, then `yarn build`.
> 
> This is what I see when I run `yarn build`:
> 
> > Charences-MacBook-Pro:frontend charencewong$ yarn build
> > yarn run v1.10.1
> > $ webpack --progress --mode production --config ./internals/webpack/webpack.prod.config.js
> > clean-webpack-plugin: /Users/charencewong/Workspace/3drepo.io/public/dist has been removed.
> > 98% after emitting swPrecacheTotal precache size is about 0 B for 0 resources.
> > ✨  Done in 180.91s.
> 
> ======
> **_EDIT:_**
> 
> I found the problem! It seems like the build process is not excluding the ~ files left behind by vi. E.g. `frontend/components/viewpoints/css/views.css~`
> 
> I assume the `test: /\.css$/` in `frontend/internals/webpack/tools/loaders.js` is supposed to filter out that kind of file, but it doesn't seem to be working :-(

Yes, this regex is correct.
Are you sure about that? Did you try to remove this files and ran build again? I've recreated this situation on my machine and the build and the app seem to work properly.

Charence commented:
> @Charence everything should work properly now. The problem was temporary solution for your legacy css files. If you are curious what I've changed, just check the last commit. ;)

Thanks @simplypixi! It works now!

I found a few problems (but they are user management ones):

![nov-25-2018 11-05-05-640p](https://user-images.githubusercontent.com/1438332/48978734-eb23d980-f0a7-11e8-9193-013748fa0cc9.gif)

Model/federation list doesn't seem to update in some cases when project name contains spaces

<img width="1680" alt="screenshot 2018-11-25 11 14 20" src="https://user-images.githubusercontent.com/1438332/48978755-6f765c80-f0a8-11e8-82e5-eeb83ec7da2e.png">

No users appear when trying to assign model/fed permissions. The list of users should be same as project permissions.

<img width="1680" alt="screenshot 2018-11-25 11 53 35" src="https://user-images.githubusercontent.com/1438332/48978773-dac02e80-f0a8-11e8-9d05-e1e26782f90c.png">

Teamspace admin icon doesn't load.

<img width="1680" alt="screenshot 2018-11-25 10 40 44" src="https://user-images.githubusercontent.com/1438332/48978781-03482880-f0a9-11e8-9fb4-a3301344cd31.png">

One request from our business guys: if the file name is a valid revision name and revision name is empty, can file name be used to populate revision name.

simplypixi commented:
> > @Charence everything should work properly now. The problem was temporary solution for your legacy css files. If you are curious what I've changed, just check the last commit. ;)
> 
> Thanks @simplypixi! It works now!
> 
> I found a few problems (but they are user management ones):
> 
> ![nov-25-2018 11-05-05-640p](https://user-images.githubusercontent.com/1438332/48978734-eb23d980-f0a7-11e8-9193-013748fa0cc9.gif)
> 
> Model/federation list doesn't seem to update in some cases when project name contains spaces
> 
> <img alt="screenshot 2018-11-25 11 14 20" width="1680" src="https://user-images.githubusercontent.com/1438332/48978755-6f765c80-f0a8-11e8-82e5-eeb83ec7da2e.png">
> 
> No users appear when trying to assign model/fed permissions. The list of users should be same as project permissions.
> 
> <img alt="screenshot 2018-11-25 11 53 35" width="1680" src="https://user-images.githubusercontent.com/1438332/48978773-dac02e80-f0a8-11e8-9d05-e1e26782f90c.png">
> 
> Teamspace admin icon doesn't load.

@Charence I've fixed admin icon.
Could you send me more info about the rest of bugs? Are any error messages in dev tools?
I can't reproduce them. :(

carmenfan commented:
@simplypixi  can you resolve the conflicts on this branch please :)

Charence commented:
> @Charence I've fixed admin icon.
> Could you send me more info about the rest of bugs? Are any error messages in dev tools?
> I can't reproduce them. :(

@simplypixi There are no relevant error messages in dev tools that I can see.

For the missing users, I noticed the problem seems to be caused by the 0px height of the element. If the height is increased, the users can be seen.

Strangely, @carmenfan was not able to reproduce this bug on her computer. We are both at the HEAD of ISSUE_1183 (a097d7). I have removed public/ and rebuilt.

![screenshot 2018-11-26 11 36 38](https://user-images.githubusercontent.com/1438332/49012210-9ef09c00-f170-11e8-9dc2-201f98d6fdb1.png)
![screenshot 2018-11-26 11 19 01](https://user-images.githubusercontent.com/1438332/49012216-a3b55000-f170-11e8-8234-885f448a8d08.png)

I will update shortly with details for the other bug.

simplypixi commented:
@carmenfan I've rebase this branch. The notifications' code is ugly. Moreover, a lot of my review comments were omitted. During the migration, we try to show you some coding path - the standard, which you should use creating new code. In some parts, the notifications are the complete denial of things we've done. You **can't** do that.

Now, I have to fix it.

@sanmont3drepo fyi

carmenfan commented:
@simplypixi Apologies! @sanmont3drepo fixed the problems I had with the branch and I got @Charence to had another look, forgetting that there's major clashes with this branch and it should've been on hold before the merge.

Regarding to coding standards, it may be helpful if you can write some pointers as to what you feel like should be followed (perhaps in an email) and we can have a discussion in house.

simplypixi commented:
> @simplypixi Apologies! @sanmont3drepo fixed the problems I had with the branch and I got @Charence to had another look, forgetting that there's major clashes with this branch and it should've been on hold before the merge.
> 
> Regarding to coding standards, it may be helpful if you can write some pointers as to what you feel like should be followed (perhaps in an email) and we can have a discussion in house.

@carmenfan 
Ok! ;) I'll prepare small docs after fixing bugs.

simplypixi commented:
@carmenfan @sanmont3drepo I think, I've done fixing notifications. Could you explain to me, what should I test this feature?

carmenfan commented:
@simplypixi Notification service utilises the socket.io connection to notify the user of things they may be interested in. At the moment, it only notifies a user when there is an issue/issues assigned to his/her job in a model.

- If a user assigns an issue to a job, users of that job who has access to the model should get a notification:
![before](https://user-images.githubusercontent.com/11945337/49146725-acd22880-f2fa-11e8-94e9-cbc36c560f99.gif)

- You can mark it as read or delete it
![before](https://user-images.githubusercontent.com/11945337/49146944-3681f600-f2fb-11e8-809b-f0a3e43c8c44.gif)

- If you click on it it will lead you to the model with the right issues filtered on the issues card:
![before](https://user-images.githubusercontent.com/11945337/49147177-db043800-f2fb-11e8-8dbb-0a60aeeb7c87.gif)

simplypixi commented:
@carmenfan I've push rebase branch, but I think that, one thing doesn't work properly. I mean:
`If a user assigns an issue to a job, users of that job who has access to the model should get a notification:`
I've tested this case using these accounts: **demo_3DRepo** and **carmen**. If I've changed something in issue **sample.6 Mark-Up** (Model: Lego Architecture), user: **carmen** didn't received any notifications. It's weird because the update request is resolved properly on frontend side.
Could you look at this?

In the meantime, I'll try to fix the bugs, which @Charence  reported before rebase.

sanmont3drepo commented:
> @carmenfan I've push rebase branch, but I think that, one thing doesn't work properly. I mean:
> `If a user assigns an issue to a job, users of that job who has access to the model should get a notification:`
> I've tested this case using these accounts: **demo_3DRepo** and **carmen**. If I've changed something in issue **sample.6 Mark-Up** (Model: Lego Architecture), user **carmen** didn't received any notifications. It's weird because the update request is resolved properly on frontend side.
> Could you look at this?
> 
> In the meantime, I'll try to fix the bugs, which @Charence reported before rebase.
 
Take in consideration two things:
1. That you run yarn install in the backend folder
2. Changing any field of the issue wont do it. It only creates the notification when you assign the issue to a job. Then users with privileges to comment and above that have that role will receive the notification
3. The user doing the assignation wont recieive the notification even if that person has the job.

I hope that helps,
cheers!

carmenfan commented:
@simplypixi 
I can't load issues from this branch. Issues API is returning an error 500
![image](https://user-images.githubusercontent.com/11945337/49220640-a6ad7c00-f3ce-11e8-95fb-e6c7a88b39e8.png)
My guess is something affected the binding? so I worked around it by importing middleware.js just before that call and the API returned fine

Once I worked around that, issues aren't being display and it seems to think I don't have permissions.
![image](https://user-images.githubusercontent.com/11945337/49220584-88478080-f3ce-11e8-8854-d9305bafd9f4.png)
left is ISSUE_1183 and right is 2.19.2 (close to head of staging but not quite), loading the same data model.

carmenfan commented:
Update: : the API error is related to lodash version upgrade. @sanmont3drepo is looking into it. But I can see issues once I downgraded the lodash version so this suggest the frontend problem is from this branch :(

carmenfan commented:
@simplypixi  if you merge staging the backend bug is fixed!

simplypixi commented:
@carmenfan @sanmont3drepo Thank you for helping me. It seems to work! ;)

simplypixi commented:
@Charence I fixed two bugs: loading models and missing admin icon. I can't reproduce situation with user permissions (wrong height). Could you send me your browser and OS version? I'll try to reflect your environment 1:1.

Charence commented:
> Could you send me your browser and OS version? I'll try to reflect your environment 1:1.

@simplypixi I'm using Google Chrome Version 70.0.3538.110 (64-bit) and macOS Sierra Version 10.12.6 (16G1618)

carmenfan commented:
@simplypixi 

I can view the issues in a model, but the add button is disabled and it claims i have no permissions to add an issue, even though I am an admin of the teamspace:
![image](https://user-images.githubusercontent.com/11945337/49287480-8694c000-f495-11e8-9d0c-7cce7e5241ee.png)

I can edit existing issues fine though.

carmenfan commented:
Some other bugs (will edit this post if i find more):

- if I go to example.org/carmen, I can see the URL gets redirected to example.org, but it gets stuck.
![before](https://user-images.githubusercontent.com/11945337/49288030-77af0d00-f497-11e8-9859-ce02fa47371d.gif)

simplypixi commented:
> Some other bugs (will edit this post if i find more):
> 
> * if I go to example.org/carmen, I can see the URL gets redirected to example.org, but it gets stuck.
>   ![before](https://user-images.githubusercontent.com/11945337/49288030-77af0d00-f497-11e8-9859-ce02fa47371d.gif)

Which solution do you prefer?
1. Redirecting to login or dashboard depending on user session state
2. Redirecting to new 404 page
:)

carmenfan commented:
@simplypixi 1 sounds good!

simplypixi commented:
@carmenfan I've added a handler for unknown routes and I've fixed permissions' bug. So, the PR is ready for the last re-test (I hope so!). 🤞
